### PR TITLE
Some improvements to thread names

### PR DESCRIPTION
**Pull request by alexrp**
**Do not merge.**

This makes debugging a little easier since GDB can actually tell you which threads are which. Should also work for ~most threads on Mac now.

### CODE REVIEW

File: mono/metadata/appdomain.c [hunk 1]
```diff
@@ -2431,6 +2431,7 @@ unload_thread_main (void *arg)
 	/* Force it to be attached to avoid racing during shutdown. */
 	thread = mono_thread_attach_full (mono_get_root_domain (), TRUE, &error);
 	mono_error_raise_exception (&error); /* FIXME don't raise here */
+	mono_thread_set_name_internal (thread->internal_thread, mono_string_new (mono_domain_get (), "Domain unloader"), TRUE);
```
Comment (kumpera) [inline mono/metadata/appdomain.c]: This is wrong, the thead is attached to the root domain so the name must be allocated there too.

File: mono/metadata/attach.c [hunk 2]
```diff
@@ -490,11 +492,12 @@ receiver_thread (void *arg)
 
 		printf ("attach: Connected.\n");
 
-		mono_thread_attach (mono_get_root_domain ());
+		MonoThread *thread = mono_thread_attach (mono_get_root_domain ());
+		mono_thread_set_name_internal (thread->internal_thread, mono_string_new (mono_domain_get (), "Attach receiver"), TRUE);
```
Comment (kumpera) [inline mono/metadata/attach.c]: Same issue here with where the thread name is allocated.

File: mono/metadata/attach.c [hunk 3]
```diff
@@ -490,11 +492,12 @@ receiver_thread (void *arg)
...
 		/* Ask the runtime to not abort this thread */
 		//mono_thread_current ()->flags |= MONO_THREAD_FLAG_DONT_MANAGE;
 		/* Ask the runtime to not wait for this thread */
-		mono_thread_internal_current ()->state |= ThreadState_Background;
+		thread->internal_thread->state |= ThreadState_Background;
```
Comment (kumpera) [inline mono/metadata/attach.c]: Are you sure it won't be null at this point? The internal thread is not the same thing as what mono_thread_attach returns.

File: mono/metadata/gc.c [hunk 4]
```diff
@@ -748,6 +748,8 @@ finalize_domain_objects (DomainFinalizationReq *req)
 static guint32
 finalizer_thread (gpointer unused)
 {
+	ves_icall_System_Threading_Thread_SetName_internal (mono_thread_current ()->internal_thread, mono_string_new (mono_domain_get (), "Finalizer"));
```
Comment (kumpera) [inline mono/metadata/gc.c]: Why are we not using mono_thread_set_name_internal ?

File: mono/mini/aot-compiler.c [hunk 5]
```diff
@@ -7865,7 +7865,8 @@ compile_thread_main (gpointer *user_data)
 	GPtrArray *methods = (GPtrArray *)user_data [2];
 	int i;
 
-	mono_thread_attach (domain);
+	MonoThread *thread = mono_thread_attach (domain);
+	mono_thread_set_name_internal (thread->internal_thread, mono_string_new (mono_domain_get (), "AOT compiler"), TRUE);
```
Comment (kumpera) [inline mono/mini/aot-compiler.c]: This won't work for cross compilation as the GC is disabled.

File: mono/mini/debugger-agent.c [hunk 6]
```diff
@@ -9690,8 +9690,10 @@ debugger_thread (void *arg)
 	debugger_thread_id = mono_native_thread_id_get ();
 
 	attach_cookie = mono_jit_thread_attach (mono_get_root_domain (), &attach_dummy);
+	MonoInternalThread *thread = mono_thread_internal_current ();
+	mono_thread_set_name_internal (thread, mono_string_new (mono_domain_get (), "Debugger agent"), TRUE);
```
Comment (kumpera) [inline mono/mini/debugger-agent.c]: Use the root domain here again.

**Comment by kumpera** [comment]
@monojenkins merge.

File: mono/utils/mono-threads-posix.c [hunk 7]
```diff
@@ -20,6 +20,7 @@
 #include <mono/metadata/gc-internals.h>
 
 #include <errno.h>
+#include <dlfcn.h>
```
Comment (akoeplinger) [inline mono/utils/mono-threads-posix.c]: This should be inside an ifdef, it currently breaks the MSVC build.

**Comment by alexrp** [comment]
Opening a new PR since I'm adding a bunch of extra commits.

**Pull request closed without merging**